File: experiments/exp005/adjoint.py
"""
MaoField Exp005: Adjoint Functor Implementation
=================================================

G = field's residual imbalance / energy density map.
Demand-driven practice selection via physical coupling.
Full F⇄G adjoint iteration.
"""

import numpy as np
from simulator_ac import (
    laplacian_3d_periodic, evolve_steps,
    compute_energy, compute_energy_density, compute_residual
)
from config import D, DT, DX, INTERACT_STEPS, RELAX_STEPS, TOP_K


def compute_coupling_residual(r, S_corpus_list):
    """Coupling between residual imbalance r(x) and each corpus source field.
    c_i = ∫ r(x) · S_i(x) dx  (volume integral = physical coupling energy)
    """
    couplings = np.array([float(np.sum(r * S_i)) for S_i in S_corpus_list])
    return couplings


def compute_coupling_energy(e, S_corpus_list):
    """Coupling between energy density e(x) and each corpus source field.
    c_i = ∫ e(x) · S_i(x) dx
    High energy regions coupling with source = where the field "needs" this text.
    """
    couplings = np.array([float(np.sum(e * S_i)) for S_i in S_corpus_list])
    return couplings


def select_practice_targets(couplings, k=TOP_K):
    """Select top-k corpus items by absolute coupling strength."""
    abs_c = np.abs(couplings)
    if k >= len(couplings):
        return np.argsort(abs_c)[::-1]
    indices = np.argsort(abs_c)[-k:][::-1]
    return indices


def practice_with_targets(u, S_query, target_indices, S_corpus_list,
                          interact_steps=INTERACT_STEPS,
                          relax_steps=RELAX_STEPS):
    """Interact with selected corpus texts sequentially, then relax.
    Relaxation uses S_query (query context stays; corpus removed).
    """
    for idx in target_indices:
        S_i = S_corpus_list[idx]
        S_combined = S_query + S_i
        mx = max(abs(S_combined.max()), abs(S_combined.min()))
        if mx > 0:
            S_combined = S_combined / mx
        # Interact
        u = evolve_steps(u, S_combined, D, DT, DX, interact_steps)
        # Relax: remove corpus, keep query context
        u = evolve_steps(u, S_query, D, DT, DX, relax_steps)
    return u


def adjoint_iteration(u, S_query, S_corpus_list, corpus_texts=None,
                      max_rounds=5, convergence_tol=0.01, k=TOP_K,
                      coupling_mode='energy'):
    """Full F⇄G adjoint iteration.

    Args:
        u: initial field (after init_steps evolution)
        S_query: query source field
        S_corpus_list: list of precomputed corpus source fields
        corpus_texts: optional, for logging
        max_rounds: max iteration rounds
        convergence_tol: stop if relative change in metric < this
        k: number of practice targets per round
        coupling_mode: 'energy' (default, safer) or 'residual'

    Returns:
        u: field after adjoint iteration
        history: list of dicts with diagnostics per round
    """
    history = []
    prev_metric = None

    for round_idx in range(max_rounds):
        # === G: compute imbalance ===
        r = compute_residual(u, S_query, D, DX)
        r_norm = float(np.sqrt(np.sum(r**2)))

        e = compute_energy_density(u, S_query, D, DX)
        e_total = float(np.sum(e))
        e_std = float(np.std(e))

        # Coupling computation
        if coupling_mode == 'residual':
            couplings = compute_coupling_residual(r, S_corpus_list)
            metric = r_norm
        else:  # 'energy'
            couplings = compute_coupling_energy(e, S_corpus_list)
            metric = e_std  # energy density spread

        # Select practice targets
        targets = select_practice_targets(couplings, k=k)

        # Diagnostics
        round_info = {
            'round': round_idx,
            'r_norm': r_norm,
            'e_total': e_total,
            'e_std': e_std,
            'coupling_mode': coupling_mode,
            'coupling_max': float(np.max(np.abs(couplings))),
            'coupling_min_selected': float(np.abs(couplings[targets[-1]])) if len(targets) > 0 else 0,
            'coupling_mean_all': float(np.mean(np.abs(couplings))),
            'selected_indices': targets.tolist(),
            'selected_couplings': [float(couplings[i]) for i in targets],
            'metric': float(metric),
        }
        if corpus_texts is not None:
            round_info['selected_texts'] = [corpus_texts[i][:30] for i in targets]

        history.append(round_info)
        print(f"    Round {round_idx}: ‖r‖={r_norm:.4f} e_std={e_std:.4f} "
              f"coupling_max={round_info['coupling_max']:.4f} "
              f"targets={len(targets)}")

        # Convergence check
        if prev_metric is not None:
            change = abs(metric - prev_metric) / (abs(prev_metric) + 1e-10)
            print(f"      metric change: {change:.4f}")
            if change < convergence_tol:
                print(f"      Converged at round {round_idx}.")
                break
        prev_metric = metric

        # === F: practice with selected corpus ===
        u = practice_with_targets(u, S_query, targets, S_corpus_list)

    return u, history


def blind_practice(u, S_query, S_corpus_list, indices,
                   interact_steps=INTERACT_STEPS,
                   relax_steps=RELAX_STEPS):
    """Blind practice: interact with given indices in order (no G selection).
    For condition B comparison.
    """
    for i, idx in enumerate(indices):
        S_i = S_corpus_list[idx]
        S_combined = S_query + S_i
        mx = max(abs(S_combined.max()), abs(S_combined.min()))
        if mx > 0:
            S_combined = S_combined / mx
        u = evolve_steps(u, S_combined, D, DT, DX, interact_steps)
        u = evolve_steps(u, S_query, D, DT, DX, relax_steps)
        if (i + 1) % 10 == 0:
            e = compute_energy(u, S_query, D, DX)
            print(f"      Blind practice {i+1}/{len(indices)}: E={e:.0f}")
    return u

File: experiments/exp006/rust_solver/src/adjoint.rs
//! Multi-component adjoint iteration: G (energy density coupling) + F (practice).

use rayon::prelude::*;
use crate::config::*;
use crate::field::{McField, Field3D, Lambdas, init_field_mc, evolve_mc, energy_density_mc, idx};

/// Compute coupling c_total = sum_i |integral e_i * S_i_corpus| for each corpus entry.
pub fn compute_couplings(e_list: &[Field3D], corpus_mc: &[Vec<Field3D>]) -> Vec<f64> {
    corpus_mc
        .par_iter()
        .map(|s_entry| {
            let mut c_total = 0.0f64;
            for (i, e_i) in e_list.iter().enumerate() {
                let c_i: f64 = e_i.iter().zip(s_entry[i].iter()).map(|(a, b)| a * b).sum();
                c_total += c_i.abs();
            }
            c_total
        })
        .collect()
}

/// Select top-k indices by coupling strength.
fn select_topk(couplings: &[f64], k: usize) -> Vec<usize> {
    let mut indexed: Vec<(usize, f64)> = couplings.iter().copied().enumerate().collect();
    indexed.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
    indexed.iter().take(k.min(indexed.len())).map(|(i, _)| *i).collect()
}

/// Practice with selected corpus entries (interact + relax per entry).
fn practice_with_targets(
    u_list: &mut McField,
    s_query: &[Field3D],
    targets: &[usize],
    corpus_mc: &[Vec<Field3D>],
    lambdas: &Lambdas,
) {
    for &idx in targets {
        // Combined source = query + corpus, per component, normalized
        let s_combined: Vec<Field3D> = (0..u_list.len())
            .map(|c| {
                let mut sc: Field3D = s_query[c]
                    .iter()
                    .zip(corpus_mc[idx][c].iter())
                    .map(|(a, b)| a + b)
                    .collect();
                let mx = sc.iter().map(|v| v.abs()).fold(0.0f64, f64::max);
                if mx > 0.0 {
                    for v in sc.iter_mut() {
                        *v /= mx;
                    }
                }
                sc
            })
            .collect();

        // Interact
        evolve_mc(u_list, &s_combined, lambdas, INTERACT_STEPS);
        // Relax with query only
        evolve_mc(u_list, s_query, lambdas, RELAX_STEPS);
    }
}

/// Round diagnostics.
#[derive(Clone, serde::Serialize)]
pub struct RoundInfo {
    pub round: usize,
    pub r_norm: f64,
    pub e_std: f64,
    pub coupling_max: f64,
    pub n_targets: usize,
}

/// Full adjoint iteration for a single text's field.
pub fn adjoint_iteration(
    u_list: &mut McField,
    s_query: &[Field3D],
    corpus_mc: &[Vec<Field3D>],
    lambdas: &Lambdas,
    label: &str,
) -> Vec<RoundInfo> {
    let mut history = Vec::new();
    let mut prev_metric: Option<f64> = None;

    for round in 0..MAX_ROUNDS {
        // G: energy density per component
        let e_list = energy_density_mc(u_list, s_query);

        // Residual norm (for diagnostics)
        let r_norm = residual_norm_mc(u_list, s_query, lambdas);

        // e_std = mean of per-component std
        let e_std = e_list
            .iter()
            .map(|e| {
                let mean = e.iter().sum::<f64>() / NCELLS as f64;
                let var = e.iter().map(|v| (v - mean) * (v - mean)).sum::<f64>() / NCELLS as f64;
                var.sqrt()
            })
            .sum::<f64>()
            / e_list.len() as f64;

        let couplings = compute_couplings(&e_list, corpus_mc);
        let targets = select_topk(&couplings, TOP_K);
        let coupling_max = couplings.iter().copied().fold(0.0f64, f64::max);

        let info = RoundInfo {
            round,
            r_norm,
            e_std,
            coupling_max,
            n_targets: targets.len(),
        };
        eprintln!(
            "    Round {}: ‖r‖={:.4} e_std={:.4} c_max={:.4} targets={}",
            round, r_norm, e_std, coupling_max, targets.len()
        );
        history.push(info);

        // Convergence check
        let metric = e_std;
        if let Some(prev) = prev_metric {
            let change = (metric - prev).abs() / (prev.abs() + 1e-10);
            eprintln!("      metric change: {:.4}", change);
            if change < CONVERGENCE_TOL {
                eprintln!("      Converged at round {}.", round);
                break;
            }
        }
        prev_metric = Some(metric);

        // F: practice
        practice_with_targets(u_list, s_query, &targets, corpus_mc, lambdas);
    }
    let _ = label; // used only for future logging
    history
}

/// Residual norm: sqrt(sum over components of sum r_i^2).
fn residual_norm_mc(u_list: &McField, s_list: &[Field3D], lambdas: &Lambdas) -> f64 {
    let n_comp = u_list.len();
    let inv_dx2 = 1.0 / (DX * DX);
    let mut total = 0.0f64;

    for i in 0..n_comp {
        let u = &u_list[i];
        let s = &s_list[i];
        let sum: f64 = (0..NCELLS)
            .into_par_iter()
            .map(|k| {
                let x = k % GRID;
                let y = (k / GRID) % GRID;
                let z = k / (GRID * GRID);
                let xm = if x == 0 { GRID - 1 } else { x - 1 };
                let xp = if x == GRID - 1 { 0 } else { x + 1 };
                let ym = if y == 0 { GRID - 1 } else { y - 1 };
                let yp = if y == GRID - 1 { 0 } else { y + 1 };
                let zm = if z == 0 { GRID - 1 } else { z - 1 };
                let zp = if z == GRID - 1 { 0 } else { z + 1 };
                let lap = (u[idx(xp, y, z)] + u[idx(xm, y, z)]
                    + u[idx(x, yp, z)] + u[idx(x, ym, z)]
                    + u[idx(x, y, zp)] + u[idx(x, y, zm)]
                    - 6.0 * u[k])
                    * inv_dx2;
                let ui = u[k];
                let mut cross = 0.0f64;
                for j in 0..n_comp {
                    if j != i {
                        let lam = lambdas.get(i, j);
                        if lam != 0.0 {
                            cross += lam * u_list[j][k];
                        }
                    }
                }
                let r = D_COEFF * lap - ui * ui * ui + ui - cross
                    + (1.0 - 3.0 * ui * ui) * s[k];
                r * r
            })
            .sum();
        total += sum;
    }
    total.sqrt()
}
